src/plugins/search/search.js

```diff
@@ -1,35 +1,152 @@
-import Hooks from './../../pluginHooks';
-import {addClass, removeClass} from './../../helpers/dom/element';
-import {registerRenderer, getRenderer} from './../../renderers';
+import BasePlugin from './../_base';
+import {registerPlugin} from './../../plugins';
+import {isObject} from './../../helpers/object';
+import {rangeEach} from './../../helpers/number';
+import {isUndefined} from './../../helpers/mixed';
+
+const DEFAULT_SEARCH_RESULT_CLASS = 'htSearchResult';
+
+const DEFAULT_CALLBACK = function(instance, row, col, data, testResult) {
+  instance.getCellMeta(row, col).isSearchResult = testResult;
+};
+
+const DEFAULT_QUERY_METHOD = function(query, value) {
+  if (isUndefined(query) || query === null || !query.toLowerCase || query.length === 0) {
+    return false;
+  }
+  if (isUndefined(value) || value === null) {
+    return false;
+  }
+
+  return value.toString().toLowerCase().indexOf(query.toLowerCase()) !== -1;
+};
 
 /**
- * @private
  * @plugin Search
+ *
+ * @example
+ *
+ * ```js
+ * ...
+ *  // as boolean
+ *  search: true
+ *
+ *  // as a object with one or more options
+ *  search: {
+ *    callback: myNewCallbackFunction,
+ *    queryMethod: myNewQueryMethod,
+ *    searchResultClass: 'customClass'
+ *  }
+ *
+ * // Access to search plugin instance:
+ * var searchPlugin = hot.getPlugin('search');
+ *
+ * // Set callback programmatically:
+ * searchPlugin.setCallback(myNewCallbackFunction);
+ * // Set query method programmatically:
+ * searchPlugin.setQueryMethod(myNewQueryMethod);
+ * // Set search result cells class programmatically:
+ * searchPlugin.setSearchResultClass(customClass);
+ * ...
+ * ```
  */
-function Search(instance) {
-  this.query = function(queryStr, callback, queryMethod) {
-    var rowCount = instance.countRows();
-    var colCount = instance.countCols();
-    var queryResult = [];
-
-    if (!callback) {
-      callback = Search.global.getDefaultCallback();
-    }
+class Search extends BasePlugin {
+  constructor(hotInstance) {
+    super(hotInstance);
+    /**
+     * Callback function is responsible for marking the cells belonging to the search by setting property to it.
+     *
+     * @type {Function}
+     */
+    this.callback = DEFAULT_CALLBACK;
+    /**
+     * Query function is responsible for determining whether a query matches the value stored in a cell.
+     *
+     * @type {Function}
+     */
+    this.queryMethod = DEFAULT_QUERY_METHOD;
+    /**
+     *  Class added to each cell that belongs to the search.
```

Comment:
Maybe change to: `Class added to each cell that belongs to the searched query.`

Reply:
Done.